zookeeper-api/src/main/java/org/apache/helix/zookeeper/util/HttpRoutingDataReader.java

```diff
@@ -0,0 +1,141 @@
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.DefaultBackoffStrategy;
+import org.apache.http.impl.client.DefaultHttpRequestRetryHandler;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+
+
+public class HttpRoutingDataReader {
+  private static final String MSDS_ENDPOINT =
+      System.getProperty(MetadataStoreRoutingConstants.MSDS_SERVER_ENDPOINT_KEY);
+
+  /** Double-checked locking requires that the following fields be volatile */
+  private static volatile Map<String, List<String>> _rawRoutingData;
+  private static volatile MetadataStoreRoutingData _metadataStoreRoutingData;
+
+  /**
+   * This class is a Singleton.
+   */
+  private HttpRoutingDataReader() {
+  }
+
+  /**
+   * Fetches routing data from the data source via HTTP.
+   * @return a mapping from "metadata store realm addresses" to lists of
+   * "metadata store sharding keys", where the sharding keys in a value list all route to
+   * the realm address in the key disallows a meaningful mapping to be returned
+   */
+  public static Map<String, List<String>> getRawRoutingData()
+      throws IOException {
+    if (MSDS_ENDPOINT == null || MSDS_ENDPOINT.isEmpty()) {
+      throw new IllegalStateException(
+          "HttpRoutingDataReader was unable to find a valid MSDS endpoint String in System Properties!");
+    }
+    if (_rawRoutingData == null) {
+      synchronized (HttpRoutingDataReader.class) {
+        if (_rawRoutingData == null) {
+          CloseableHttpResponse routingDataResponse = getAllRoutingData();
+          // Update the reference if reading routingData over HTTP is successful
+          _rawRoutingData = parseRoutingData(routingDataResponse);
+        }
+      }
+    }
+    return _rawRoutingData;
+  }
+
+  /**
+   * Returns the routing data read from MSDS in a MetadataStoreRoutingData format.
+   * @return
+   * @throws IOException if there is an issue connecting to MSDS
+   * @throws InvalidRoutingDataException if the raw routing data is not valid
+   */
+  public static MetadataStoreRoutingData getMetadataStoreRoutingData()
+      throws IOException, InvalidRoutingDataException {
+    if (_metadataStoreRoutingData == null) {
+      synchronized (HttpRoutingDataReader.class) {
+        if (_metadataStoreRoutingData == null) {
+          _metadataStoreRoutingData = new TrieRoutingData(getRawRoutingData());
+        }
+      }
+    }
+    return _metadataStoreRoutingData;
+  }
+
+  /**
+   * Makes an HTTP call to fetch all routing data.
+   * @return
+   * @throws IOException
+   */
+  private static CloseableHttpResponse getAllRoutingData()
+      throws IOException {
+    // Retry count is 3 by default
+    HttpGet requestAllData = new HttpGet(
+        MSDS_ENDPOINT + MetadataStoreRoutingConstants.MSDS_GET_ALL_ROUTING_DATA_ENDPOINT);
+    try (CloseableHttpClient httpClient = HttpClients.custom()
+        .setConnectionBackoffStrategy(new DefaultBackoffStrategy())
```

Comment:
I don't see a timeout setting. If we don't set timeout, it may use system defined timeout or infinite timeout. To make the system responsive and stable, are we considering timeout: connectTimeout, socketTimeout, etc.?

Reply:
Set.